woocommerce/woocommerce
# Split the `can create product, attributes and variations, edit variations and delete variations` E2E test into smaller tests

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/37724.

The major change in this PR is the breaking down of the `can create product, attributes and variations, edit variations and delete variations` test into smaller tests, like so:

   <img width="399" alt="TO0LUpboUk" src="https://user-images.githubusercontent.com/4509348/232119378-e1787318-e825-4db4-b2f9-93df3401adea.png">

There are also these changes:

1.  Added `test.step()` to (hopefully) improve readability within the spec file, and in the Playwright or Allure HTML reports.

    With this change, the steps that used to look like this:

    <img width="422" alt="BprEHt2BsL" src="https://user-images.githubusercontent.com/4509348/232119255-7650bc80-56ba-4609-b576-c9ca51da60de.png">

    ...would now look like this:

    <img width="531" alt="OqI5ZyaYZl" src="https://user-images.githubusercontent.com/4509348/232119324-b55531b8-af38-44f1-9e6c-62d43516e781.png">

    However to limit the scope of this PR, it does not yet add `test.step()`'s to the `can manually add a variation, manage stock levels, set variation defaults and remove a variation` test.

1.  We need to have a way to make each of the smaller tests independent from one another. That's why the function [`createVariableProductFixture`](https://github.com/woocommerce/woocommerce/blob/e2e/split-variable-product-test/plugins/woocommerce/tests/e2e-pw/tests/merchant/create-variable-product.spec.js#L86) was created. It creates a fixture that will be used by each of the smaller tests, so they would no longer have to rely on the first test for creating the variable product.
1.  The resetting of the variable product tour now [happens at the beginning of the very first test](https://github.com/woocommerce/woocommerce/blob/e2e/split-variable-product-test/plugins/woocommerce/tests/e2e-pw/tests/merchant/create-variable-product.spec.js#L224-L229) (`can create product, attributes and variations`), instead of in the `afterAll` hook. This is due to the fact that only the first test relies on the tour to be displayed. The other tests do not.
1.  Updated some of the locators to using `page.getByRole()` for better readability.
1.  Added assertions to tests that turned out to be not having any after the split.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Assuming that your local environment is already running, and that you're in the `plugins/woocommerce` directory:

1. Run the tests individually to verify that they're not interdependent with each other. They should all pass:
   - `pnpm test:e2e-pw -g "can create product, attributes and variations"`
   - `pnpm test:e2e-pw -g "can individually edit variations"`
   - `pnpm test:e2e-pw -g "can bulk edit variations"`
   - `pnpm test:e2e-pw -g "can delete all variations"`
   - `pnpm test:e2e-pw -g "can manually add a variation, manage stock levels, set variation defaults and remove a variation"`
2. Run the entire `create-variable-product.spec.js` spec. It should pass.
   - `pnpm test:e2e-pw create-variable-product.spec.js`
